README.md
# shelf_gzip

[![pub package](https://img.shields.io/pub/v/shelf_gzip.svg?logo=dart&logoColor=00b9fc)](https://pub.dev/packages/shelf_gzip)
[![Null Safety](https://img.shields.io/badge/null-safety-brightgreen)](https://dart.dev/null-safety)
[![Dart CI](https://github.com/johnpryan/shelf_gzip/actions/workflows/dart.yml/badge.svg)](https://github.com/johnpryan/shelf_gzip/actions/workflows/dart.yml)
[![GitHub Tag](https://img.shields.io/github/v/tag/johnpryan/shelf_gzip?logo=git&logoColor=white)](https://github.com/johnpryan/shelf_gzip/releases)
[![New Commits](https://img.shields.io/github/commits-since/johnpryan/shelf_gzip/latest?logo=git&logoColor=white)](https://github.com/johnpryan/shelf_gzip/network)
[![Last Commits](https://img.shields.io/github/last-commit/johnpryan/shelf_gzip?logo=git&logoColor=white)](https://github.com/johnpryan/shelf_gzip/commits/master)
[![Pull Requests](https://img.shields.io/github/issues-pr/johnpryan/shelf_gzip?logo=github&logoColor=white)](https://github.com/johnpryan/shelf_gzip/pulls)
[![Code size](https://img.shields.io/github/languages/code-size/johnpryan/shelf_gzip?logo=github&logoColor=white)](https://github.com/johnpryan/shelf_gzip)
[![License](https://img.shields.io/github/license/johnpryan/shelf_gzip?logo=open-source-initiative&logoColor=green)](https://github.com/johnpryan/shelf_gzip/blob/master/LICENSE)

Shelf middleware to GZIP encoding responses, with compression level and compression scope by content-type.

## Usage

```dart
import 'dart:async' show runZonedGuarded;
import 'dart:io';

import 'package:path/path.dart' show join, dirname;
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf/shelf_io.dart' as shelf_io;
import 'package:shelf_gzip/shelf_gzip.dart';
import 'package:shelf_static/shelf_static.dart';

void main() {
  // Assumes the server lives in bin/ and that `pub build` ran
  var pathToBuild =
  join(dirname(Platform.script.toFilePath()), '..', 'build/web');
  var staticHandler =
  createStaticHandler(pathToBuild, defaultDocument: 'index.html');

  var portEnv = Platform.environment['PORT'];
  var port = portEnv == null ? 9999 : int.parse(portEnv);

  runZonedGuarded(() async {
    var handler = const shelf.Pipeline()
            .addMiddleware(gzipMiddleware)
            .addHandler(staticHandler);

    await shelf_io.serve(handler, '0.0.0.0', port);

    print("Serving $pathToBuild on port $port");
  }, (e, stackTrace) => print('Server error: $e $stackTrace'));
}
```

## When not compress 

The `gzip` encoding won't be applied if:

- The `Content-Type` is for an already
compressed type. See `isAlreadyCompressedContentType`.


- A small response body (length < 512) will not benefit from being compressed.

## Compression Level

The default `gzip` encoder compression level is set to **4**,
since this is the recommended level for live compression
(not stored file compression). A level **4** compression
has the best trade off for text/code content and CPU usage.

- *The original default `gzip` encoder compression level is **6**.*

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/johnpryan/shelf_gzip

# Authors
- John Ryan: [johnpryan][github_johnpryan].
- Graciliano M. Passos: [gmpassos][github_gmpassos].

[github_johnpryan]: https://github.com/johnpryan
[github_gmpassos]: https://github.com/gmpassos


## License

[BSD-3-Clause License][license]

[license]: https://github.com/johnpryan/shelf_gzip/blob/master/LICENSE


CHANGELOG.md
## 4.1.0

- `createGzipMiddleware` and `gzipEncodeResponse`:
  - Added parameters `addServerTiming` and `serverTimingEntryName`. 

- lints: ^1.0.1

## 4.0.1

- `X-Compression-Ratio`: truncate ratio precision `> 4`.
- Update `.gitignore`

## 4.0.0

- Null Safety compliant.
- Added:
  - `createGzipMiddleware`: can create a `Middleware` with a custom `compressionLevel`.
  - `acceptsGzipEncoding`: checks if a `Request` accepts `gzip` encoding.
  - `gzipEncodeResponse`: converts a `Response` to a `gzip` encoding response.
  - `isAlreadyCompressedContentType` and `isAlreadyCompressedExtension`: checks if is already compressed.
- Optimized bytes reading and compression.
- Added an example.
- Improved API documentation.
- lints: ^1.0.0
  - Using `lints/recommended.yaml`
- test: ^1.16.0
  - Added basic tests.
- CI: Added GitHub action.

## 1.0.0

- Initial version


test/shelf_gzip_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:shelf/shelf.dart';
import 'package:shelf/shelf_io.dart';
import 'package:shelf_gzip/shelf_gzip.dart';
import 'package:test/test.dart';

final Uint8List _pngTransparentPixel = base64.decode(
    'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==');

void main() {
  group('shelf_gzip', () {
    test('basic', () async {
      // `shelf` Pipeline:
      var pipeline = const Pipeline()
          .addMiddleware(logRequests())
          .addMiddleware(gzipMiddleware);

      var handler = pipeline.addHandler((request) {
        var requestedUri = request.requestedUri;
        var path = requestedUri.path;
        if (path == '/transparent-pixel') {
          print('Sending PNG... $requestedUri');
          return Response.ok(_pngTransparentPixel,
              headers: {'Content-Type': 'image/png'});
        } else {
          return Response.ok('Requested: $requestedUri');
        }
      });

      var server = await serve(handler, 'localhost', 0);

      var port = server.port;

      var baseURL = 'http://localhost:$port';

      print('baseURL: $baseURL');

      // No gzip, small body:
      expect(
          await _getURL('$baseURL/foo',
              expectedContentEncoding: '',
              compressionState:
                  HttpClientResponseCompressionState.notCompressed),
          equals('Requested: http://localhost:$port/foo'));

      var longValue = 'long_value_'.padRight(1024, 'x');

      // Expected gzip, long body:
      expect(
          await _getURL('$baseURL/$longValue',
              expectedContentEncoding: 'gzip',
              compressionState:
                  HttpClientResponseCompressionState.decompressed),
          equals('Requested: http://localhost:$port/$longValue'));

      // Expected a PNG without gzip encoding:
      expect(
          await _getURL('$baseURL/transparent-pixel',
              expectedContentType: 'image/png',
              expectedContentEncoding: '',
              compressionState:
                  HttpClientResponseCompressionState.notCompressed),
          startsWith('\u0089PNG\r\n'));

      server.close();
    });

    test('Check added headers', () async {
      // `shelf` Pipeline:
      var pipeline = const Pipeline()
          .addMiddleware(logRequests())
          .addMiddleware(createGzipMiddleware(
            addCompressionRatioHeader: true,
            addServerTiming: true,
            serverTimingEntryName: 'x-gzip',
          ));

      var handler = pipeline.addHandler((request) {
        var requestedUri = request.requestedUri;
        return Response.ok('Requested: $requestedUri');
      });

      var server = await serve(handler, 'localhost', 0);

      var port = server.port;

      var baseURL = 'http://localhost:$port';

      print('baseURL: $baseURL');

      // No gzip, small body:
      expect(
          await _getURL(
            '$baseURL/foo',
            expectedContentEncoding: '',
            compressionState: HttpClientResponseCompressionState.notCompressed,
            expectedHeaders: {
              'x-compression-ratio': isNull,
              'server-timing': isNull,
            },
          ),
          equals('Requested: http://localhost:$port/foo'));

      var longValue = 'long_value_'.padRight(1024, 'x');

      // Expected gzip, long body:
      expect(
          await _getURL(
            '$baseURL/$longValue',
            expectedContentEncoding: 'gzip',
            compressionState: HttpClientResponseCompressionState.decompressed,
            expectedHeaders: {
              'x-compression-ratio': isNotEmpty,
              'server-timing': startsWith('x-gzip;dur='),
            },
          ),
          equals('Requested: http://localhost:$port/$longValue'));

      server.close();
    });

    test('isAlreadyCompressedExtension', () async {
      expect(isAlreadyCompressedExtension('gz'), isTrue);
      expect(isAlreadyCompressedExtension('gzip'), isTrue);
      expect(isAlreadyCompressedExtension('zip'), isTrue);

      expect(isAlreadyCompressedExtension('png'), isTrue);
      expect(isAlreadyCompressedExtension('jpeg'), isTrue);
      expect(isAlreadyCompressedExtension('jpg'), isTrue);

      expect(isAlreadyCompressedExtension('txt'), isFalse);
      expect(isAlreadyCompressedExtension('text'), isFalse);
      expect(isAlreadyCompressedExtension('html'), isFalse);
    });

    test('isAlreadyCompressedContentType', () async {
      expect(isAlreadyCompressedContentType('application/gzip'), isTrue);
      expect(isAlreadyCompressedContentType('application/zip'), isTrue);

      expect(isAlreadyCompressedContentType('image/png'), isTrue);
      expect(isAlreadyCompressedContentType('image/jpeg'), isTrue);

      expect(isAlreadyCompressedContentType('text/plain'), isFalse);
      expect(isAlreadyCompressedContentType('text/html'), isFalse);

      expect(isAlreadyCompressedContentType('application/json'), isFalse);
      expect(isAlreadyCompressedContentType('application/javascript'), isFalse);

      expect(isAlreadyCompressedContentType('png'), isTrue);
      expect(isAlreadyCompressedContentType('json'), isFalse);
    });
  });
}

/// Simple HTTP get URL function.
Future<String> _getURL(String url,
    {Map<String, dynamic>? parameters,
    String? expectedContentType,
    String? expectedContentEncoding,
    HttpClientResponseCompressionState? compressionState,
    Map<String, dynamic>? expectedHeaders}) async {
  var uri = Uri.parse(url);

  if (parameters != null) {
    parameters = parameters.map((key, value) => MapEntry(key, '$value'));

    uri = Uri(
      scheme: uri.scheme,
      userInfo: uri.userInfo,
      host: uri.host,
      port: uri.port,
      path: uri.path,
      fragment: uri.fragment,
      queryParameters: parameters,
    );
  }

  var httpClient = HttpClient();

  var response =
      await httpClient.getUrl(uri).then((request) => request.close());

  var headerContentType = response.headers[HttpHeaders.contentTypeHeader];
  var headerContentEncoding =
      response.headers[HttpHeaders.contentEncodingHeader];

  var headerCompressionRatio = response.headers['X-Compression-Ratio'];

  print('Client> '
      'Content-Type: $headerContentType ; '
      'Content-Encoding: $headerContentEncoding ; '
      'X-Compression-Ratio: $headerCompressionRatio ; '
      'compressionState: ${response.compressionState} '
      '> $url');

  if (expectedContentType != null) {
    var contentType = headerContentType ?? [''];
    expect(contentType.first, equals(expectedContentType));
  }

  if (expectedContentEncoding != null) {
    var contentEncoding = headerContentEncoding ?? [''];
    expect(contentEncoding.first, equals(expectedContentEncoding));
  }

  if (compressionState != null) {
    expect(response.compressionState, equals(compressionState));
  }

  if (expectedHeaders != null) {
    print('HttpClientResponse Headers:');
    response.headers.forEach((k, v) => print('-- $k: $v'));

    for (var e in expectedHeaders.entries) {
      expect(response.headers.value(e.key), e.value);
    }
  }

  var data = await response.transform(Latin1Decoder()).toList();
  var body = data.join();

  return body;
}


example/shelf_gzip_example.dart
import 'dart:io';

import 'package:path/path.dart' show join, dirname;
import 'package:shelf/shelf.dart' as shelf;
import 'package:shelf/shelf_io.dart' as shelf_io;
import 'package:shelf_gzip/shelf_gzip.dart';
import 'package:shelf_static/shelf_static.dart';

void main() async {
  // Assumes the server lives in bin/ and that `pub build` ran
  var pathToBuild =
      join(dirname(Platform.script.toFilePath()), '..', 'build/web');
  var staticHandler =
      createStaticHandler(pathToBuild, defaultDocument: 'index.html');

  var portEnv = Platform.environment['PORT'];
  var port = portEnv == null ? 9999 : int.parse(portEnv);

  var handler = const shelf.Pipeline()
      .addMiddleware(gzipMiddleware) // Adds the Gzip encoding `Middleware`.
      .addHandler(staticHandler);

  await shelf_io.serve(handler, '0.0.0.0', port);

  print("Serving $pathToBuild on port $port");
}


lib/src/gzip_encoding.dart
import 'dart:async';
import 'dart:io' show HttpHeaders, ZLibEncoder;
import 'dart:math' as math;
import 'dart:typed_data';

import 'package:shelf/shelf.dart';

/// Using level 4, since have almost the same compression ratio of normal
/// texting content, but with a lower CPU usage. This is the recommended
/// for live data compression with `gzip`.
const int _defaultGzipCompressionLevel = 4;

/// A body with less than 512 bytes will have almost no benefit to be
/// compressed by `Gzip`, since a `gzip` encoded response will add an extra
/// header to the HTTP protocol and extra `Gzip` bytes to the compressed data.
const int _defaultMinimalGzipContentLength = 512;

final _defaultGzipEncoder =
    ZLibEncoder(gzip: true, level: _defaultGzipCompressionLevel);

/// The default `gzip` encoding [Middleware].
final Middleware gzipMiddleware = createGzipMiddleware();

/// Converts a [Response] to a `gzip` encoding
/// (only if the [Request] [acceptsGzipEncoding]).
///
/// - [minimalGzipContentLength] is the minimal size for a content to be
///   compressed. Default to `512`.
/// - [alreadyCompressedContentType] is a function that returns `true` if the
///   passed `contentType` is already compressed, like a `PNG`, `JPEG` or `Zip`.
///   Defaults to [isAlreadyCompressedContentType].
/// - The [compressionLevel] for the `gzip` encoder. Default: 4
/// - If [addCompressionRatioHeader] is `true`, add header `X-Compression-Ratio`.
/// - If [addServerTiming] is `true`, include or append Gzip encoding timing
///   to the `server-timing` header.
/// - [serverTimingEntryName] is the entry name to be used in
///   the `server-timing` header.
Middleware createGzipMiddleware({
  int minimalGzipContentLength = _defaultMinimalGzipContentLength,
  _AlreadyCompressedContentType? alreadyCompressedContentType,
  int compressionLevel = _defaultGzipCompressionLevel,
  bool addCompressionRatioHeader = true,
  bool addServerTiming = false,
  String serverTimingEntryName = 'gzip',
}) {
  return (Handler innerHandler) {
    return (request) {
      if (!acceptsGzipEncoding(request)) {
        return innerHandler(request);
      }
      return Future.sync(() => innerHandler(request))
          .then((response) => gzipEncodeResponse(
                response,
                minimalGzipContentLength: minimalGzipContentLength,
                alreadyCompressedContentType: alreadyCompressedContentType,
                compressionLevel: compressionLevel,
                addCompressionRatioHeader: addCompressionRatioHeader,
                addServerTiming: addServerTiming,
                serverTimingEntryName: serverTimingEntryName,
              ));
    };
  };
}

/// Returns `true` if the request accepts `gzip` encoding.
bool acceptsGzipEncoding(Request request) {
  var acceptEncoding = request.headers[HttpHeaders.acceptEncodingHeader];
  return acceptEncoding?.contains('gzip') ?? false;
}

/// Converts [response] to a `gzip` encoding response.
/// Checks [canGzipEncodeResponse].
///
/// - [minimalGzipContentLength] is the minimal size for a content to be
///   compressed. Default to `512`.
/// - [alreadyCompressedContentType] is a function that returns `true` if the
///   passed `contentType` is already compressed, like a `PNG`, `JPEG` or `Zip`.
///   Defaults to [isAlreadyCompressedContentType].
/// - If [addCompressionRatioHeader] is `true` it adds the header
///   `X-Compression-Ratio`, with compression info, e.g.: `0.55 (550/1000)`
/// - If [addServerTiming] is `true`, include or append Gzip encoding timing to
///   the `server-timing` header.
/// - [serverTimingEntryName] is the entry name to be used in the
///   `server-timing` header.
/// See [createGzipMiddleware].
FutureOr<Response> gzipEncodeResponse(
  Response response, {
  int minimalGzipContentLength = _defaultMinimalGzipContentLength,
  _AlreadyCompressedContentType? alreadyCompressedContentType,
  int compressionLevel = _defaultGzipCompressionLevel,
  bool addCompressionRatioHeader = true,
  bool addServerTiming = false,
  String serverTimingEntryName = 'gzip',
}) async {
  if (!canGzipEncodeResponse(response,
      minimalGzipContentLength: minimalGzipContentLength,
      alreadyCompressedContentType: alreadyCompressedContentType)) {
    return response;
  }

  var gzipInit = DateTime.now();

  var bufferInitialCapacity = response.contentLength ?? 1024 * 4;

  // Read the body bytes from the response:
  var bytesBuffer = await response.read().fold<_BytesBuffer>(
      _BytesBuffer(bufferInitialCapacity),
      (result, bytes) => result..addAll(bytes));

  var gzipEncoder = compressionLevel == _defaultGzipCompressionLevel
      ? _defaultGzipEncoder
      : ZLibEncoder(gzip: true, level: compressionLevel);

  // Compressed body:
  var compressedBody = gzipEncoder.convert(bytesBuffer.toUint8List());

  var bodyLength = bytesBuffer.length;
  var compressedBodyLength = compressedBody.length;

  var headers = Map<String, String>.from(response.headers);

  headers[HttpHeaders.contentEncodingHeader] = 'gzip';
  headers[HttpHeaders.contentLengthHeader] = compressedBodyLength.toString();

  if (addCompressionRatioHeader) {
    var compressionRatio = compressedBodyLength / bodyLength;

    var compressionRatioStr = '$compressionRatio';
    if (compressionRatioStr.length > 6) {
      compressionRatioStr = compressionRatio.toStringAsFixed(4);
    }

    headers['X-Compression-Ratio'] =
        '$compressionRatioStr ($compressedBodyLength/$bodyLength)';
  }

  if (addServerTiming) {
    const headerServerTiming = 'server-timing';

    var gzipTime = DateTime.now().difference(gzipInit);
    var dur = gzipTime.inMicroseconds / 1000;

    var serverTiming2 = StringBuffer();

    var serverTiming = headers[headerServerTiming];
    if (serverTiming != null && serverTiming.isNotEmpty) {
      serverTiming2.write(serverTiming);
      serverTiming2.write(',');
    }

    serverTiming2.write(serverTimingEntryName);
    serverTiming2.write(';dur=');
    serverTiming2.write(dur);

    headers[headerServerTiming] = serverTiming2.toString();
  }

  return response.change(headers: headers, body: compressedBody);
}

/// a function that returns `true` if the passed `contentType`
/// is already compressed.
typedef _AlreadyCompressedContentType = bool Function(String contentType);

/// Returns `true` if [response] can be compressed.
///
/// - [minimalGzipContentLength] is the minimal size for a content to be
///   compressed. Default to `512`.
/// - [alreadyCompressedContentType] is a function that returns `true` if the
///   passed `contentType` is already compressed, like a `PNG`, `JPEG` or `Zip`.
///
/// Checks:
/// - `Content-Encoding`: if already present, can't change it to `gzip`.
/// - `Content-Type`: checks if [isAlreadyCompressedContentType].
/// - `Content-Length`: checks if too small for compression (< 512).
bool canGzipEncodeResponse(Response response,
    {int minimalGzipContentLength = _defaultMinimalGzipContentLength,
    _AlreadyCompressedContentType? alreadyCompressedContentType}) {
  var headerContentEncoding =
      response.headers[HttpHeaders.contentEncodingHeader];

  // If the response already defines a `Content-Encoding` header it
  // won't apply the the `gzip` encoding to preserve the response behavior.
  if (headerContentEncoding != null && headerContentEncoding.isNotEmpty) {
    return false;
  }

  var responseContentLength = response.contentLength;

  // A small body will not benefit from being compressed:
  if (responseContentLength != null &&
      responseContentLength < minimalGzipContentLength) {
    return false;
  }

  var headerContentType = response.headers[HttpHeaders.contentTypeHeader];

  // Do not compress if the `Content-Type` is an already compressed type:
  if (headerContentType != null) {
    alreadyCompressedContentType ??= isAlreadyCompressedContentType;
    if (alreadyCompressedContentType(headerContentType)) {
      return false;
    }
  }

  return true;
}

/// Returns `true` if [contentType] is already compressed.
///
/// See [isAlreadyCompressedExtension].
bool isAlreadyCompressedContentType(String contentType) {
  contentType = contentType.toLowerCase().trim();
  if (contentType.isEmpty) return false;

  var idx = contentType.indexOf(';');
  if (idx >= 0) {
    contentType = contentType.substring(0, idx).trim();
  }

  var idx2 = contentType.indexOf('/');

  var type = idx2 >= 0 ? contentType.substring(idx2 + 1) : contentType;

  if (isAlreadyCompressedExtension(type)) {
    return true;
  }

  if (type.contains('+')) {
    var list = type.split('+');
    var alreadyCompressed =
        list.where((e) => isAlreadyCompressedExtension(e)).isNotEmpty;
    return alreadyCompressed;
  }

  return false;
}

bool isAlreadyCompressedExtension(String extension) {
  extension = extension.toLowerCase().trim();
  if (extension.isEmpty) return false;

  switch (extension) {
    case 'ico':
    case 'png':
    case 'jpg':
    case 'jpeg':
    case 'avi':
    case 'mp3':
    case 'mp4':
    case 'mpeg':
    case 'ogg':
    case 'ogx':
    case 'weba':
    case 'webm':
    case 'webp':
    case 'epub':
    case 'pdf':
    case 'woff':
    case 'woff2':
    case 'jar':
    case 'war':
    case '7z':
    case 'bz':
    case 'bz2':
    case 'gzip':
    case 'gz':
    case 'rar':
    case 'zip':
      return true;
    default:
      return false;
  }
}

// Optimized buffer to read the body.
class _BytesBuffer {
  Uint8List _bytes;
  int _length = 0;

  _BytesBuffer(int initialCapacity) : _bytes = Uint8List(initialCapacity);

  int get capacity => _bytes.length;

  int get length => _length;

  void _ensureCapacity(int needed) {
    if (capacity < needed) {
      var newCapacity = math.max(capacity * 2, needed);
      var bs = Uint8List(newCapacity);
      bs.addAll(_bytes);
      _bytes = bs;
    }
  }

  void addAll(List<int> bs) {
    var bsLength = bs.length;

    _ensureCapacity(_length + bsLength);

    _bytes.setAll(_length, bs);
    _length += bsLength;
  }

  Uint8List toUint8List() {
    if (_length == _bytes.length) {
      return _bytes;
    } else {
      return _bytes.sublist(0, _length);
    }
  }
}
